### PR TITLE
Verify COSE signatures

### PR DESCRIPTION
**Pull request by errfrom**

### CODE REVIEW

File: src/HydraAuctionOnchain/Helpers.hs [hunk 1]
```diff
@@ -180,31 +180,29 @@ putxoAddress = phoistAcyclic $
   plam $ \utxo ->
     pfield @"address" #$ pfield @"resolved" # utxo
 
-pvaluePaidToGeneric
-  :: Term
-      s
-      ( PTxInfo
-          :--> a
-          :--> (PAddress :--> a :--> PBool)
```
Comment (euonymos) [inline src/HydraAuctionOnchain/Helpers.hs:188]: Just curious: didn't the predicate-based version work well?

File: src/HydraAuctionOnchain/Lib/Cose.hs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+          :--> PBool
+      )
+pverifyCoseSignature = phoistAcyclic $
+  plam $ \signature vk addr payload payloadLength -> P.do
+    verifyForNetwork <- plet $ plam $ \isMainnet ->
+      pmatch (pmkSigStructure # isMainnet # addr # payload # payloadLength) $ \case
+        PNothing -> pcon PFalse
+        PJust sigStruct ->
+          pverifyEd25519Signature # vk # sigStruct # signature
+    (verifyForNetwork # pcon PFalse) -- verify for testnet
```
Comment (euonymos) [inline src/HydraAuctionOnchain/Lib/Cose.hs:55]: Can't we use `isMainnet` somehow?

File: src/HydraAuctionOnchain/Lib/Cose.hs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        addrSerializedFields <- pletFields @["addrCbor", "addrMapEntrySize"] addrSerialized
+        pjust
+          # mconcat
+            [ phexByteStr "846a5369676e61747572653158"
+            , addrSerializedFields.addrMapEntrySize
+            , phexByteStr "A201276761646472657373" -- 61646472657373 = "address"
+            , addrSerializedFields.addrCbor
+            , phexByteStr "4058"
+            , payloadLength
+            , payload
```
Comment (euonymos) [inline src/HydraAuctionOnchain/Lib/Cose.hs:35]: Does `payloadLength`have a fixed length? Look like it doesn't. Do we need to parse that signature structure?

**Review by euonymos — APPROVED** [review]
I am afraid my review cannot be super helpful to you @errfrom, but it was interesting to me. I will spend some time on CIP-8 later on. It would be great if you could find time to answer questions I left.

File: src/HydraAuctionOnchain/Lib/Cose.hs (hunk 3 above)
Comment (errfrom) [reply src/HydraAuctionOnchain/Lib/Cose.hs:35]: You are correct, the payload length is different for seller and bidder messages.

**Pull request merged**